File: alarm_node/src/utils/circuitBreaker.ts
import logger from './logger';
import { CircuitBreakerOptions, CircuitBreakerStatus } from '../types';

type CircuitState = 'CLOSED' | 'OPEN' | 'HALF_OPEN';

export class CircuitBreaker {
  private name: string;
  private state: CircuitState = 'CLOSED';
  private failureCount: number = 0;
  private successCount: number = 0;
  private lastFailureTime: number | null = null;
  private nextAttemptTime: number | null = null;
  
  private failureThreshold: number;
  private timeout: number;
  
  // Mutex to prevent race conditions during HALF_OPEN state
  private halfOpenLock: boolean = false;
  private halfOpenRequestsInFlight: number = 0;
  private halfOpenMaxConcurrent: number = 1; // Only allow 1 request at a time in HALF_OPEN

  constructor(name: string, options: CircuitBreakerOptions = {}) {
    this.name = name;
    this.failureThreshold = options.failureThreshold || 5;
    this.timeout = options.timeout || 60000;
    
    logger.info(`Circuit breaker initialized: ${this.name}`, {
      failureThreshold: this.failureThreshold,
      timeout: this.timeout
    });
  }

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    // If OPEN, check if we should try recovery (HALF_OPEN)
    if (this.state === 'OPEN') {
      const now = Date.now();
      const nextAttempt = this.nextAttemptTime || 0;
      
      if (now < nextAttempt) {
        // Still in timeout period, reject immediately but mark as retryable
        const error: any = new Error(`Circuit breaker is OPEN for ${this.name}. Will retry after ${Math.ceil((nextAttempt - now) / 1000)}s`);
        error.code = 'CIRCUIT_BREAKER_OPEN';
        error.retryable = true; // Mark as retryable so it can be retried after timeout
        error.retryAfter = Math.ceil((nextAttempt - now) / 1000); // Seconds until retry
        throw error;
      }
      
      // Timeout expired, try recovery - but only if we're not already in HALF_OPEN
      if (!this.halfOpenLock) {
        this.halfOpenLock = true;
        this.state = 'HALF_OPEN';
        this.successCount = 0; // Reset success count for recovery attempt
        this.halfOpenRequestsInFlight = 0;
        logger.info(`Circuit breaker entering HALF_OPEN state: ${this.name} - Attempting recovery`);
      }
    }
    
    // If HALF_OPEN, only allow limited concurrent requests to prevent race conditions
    if (this.state === 'HALF_OPEN') {
      if (this.halfOpenRequestsInFlight >= this.halfOpenMaxConcurrent) {
        // Too many requests in flight during HALF_OPEN, reject this one
        // This prevents the race condition where multiple requests fail the recovery
        const error: any = new Error(`Circuit breaker is in HALF_OPEN for ${this.name}, recovery in progress, please wait`);
        error.code = 'CIRCUIT_BREAKER_HALF_OPEN_BUSY';
        error.retryable = true;
        error.retryAfter = 2; // Retry in 2 seconds
        throw error;
      }
      this.halfOpenRequestsInFlight++;
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    } finally {
      if (this.state === 'HALF_OPEN' || this.halfOpenLock) {
        this.halfOpenRequestsInFlight = Math.max(0, this.halfOpenRequestsInFlight - 1);
      }
    }
  }

  private onSuccess(): void {
    this.failureCount = 0;
    
    if (this.state === 'HALF_OPEN') {
      this.successCount++;
      
      // Immediately close on first success to prevent race condition
      // This is more aggressive but prevents the constant OPEN/HALF_OPEN cycling
      if (this.successCount >= 1) { // Changed from successThreshold to 1
        this.state = 'CLOSED';
        this.successCount = 0;
        this.halfOpenLock = false;
        this.halfOpenRequestsInFlight = 0;
        logger.info(`Circuit breaker CLOSED: ${this.name} - Service recovered after ${this.successCount} success(es)`);
      }
    }
  }

  private onFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();
    
    if (this.state === 'HALF_OPEN') {
      // In HALF_OPEN, a single failure returns us to OPEN
      this.state = 'OPEN';
      this.nextAttemptTime = Date.now() + this.timeout;
      this.successCount = 0;
      this.halfOpenLock = false;
      this.halfOpenRequestsInFlight = 0;
      logger.warn(`Circuit breaker OPEN: ${this.name} - Recovery failed`);
    } else if (this.state === 'CLOSED' && this.failureCount >= this.failureThreshold) {
      this.state = 'OPEN';
      this.nextAttemptTime = Date.now() + this.timeout;
      this.successCount = 0;
      logger.error(`Circuit breaker OPEN: ${this.name} - Failure threshold reached`, {
        failureCount: this.failureCount,
        threshold: this.failureThreshold
      });
    }
  }

  reset(): void {
    this.state = 'CLOSED';
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = null;
    this.nextAttemptTime = null;
    this.halfOpenLock = false;
    this.halfOpenRequestsInFlight = 0;
    logger.info(`Circuit breaker manually reset: ${this.name}`);
  }

  getStatus(): CircuitBreakerStatus {
    return {
      name: this.name,
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
      nextAttemptTime: this.nextAttemptTime
    };
  }

  isAvailable(): boolean {
    if (this.state === 'CLOSED') {
      return true;
    }
    
    if (this.state === 'HALF_OPEN') {
      // Only available if we have room for more requests
      return this.halfOpenRequestsInFlight < this.halfOpenMaxConcurrent;
    }
    
    if (this.state === 'OPEN' && Date.now() >= (this.nextAttemptTime || 0)) {
      return true;
    }
    
    return false;
  }
}

File: metric_engine_node/engine/circuit_breaker.py
"""
Circuit Breaker for Metric Engine (DB / RabbitMQ).
Opens after N consecutive failures; recovery after timeout.
"""
import asyncio
import logging
import time
from enum import Enum
from typing import Callable, Any, Optional

logger = logging.getLogger(__name__)


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class CircuitBreaker:
    """Async circuit breaker. Opens after failure_threshold; recovery after recovery_timeout."""

    def __init__(
        self,
        failure_threshold: int = 5,
        recovery_timeout: float = 60.0,
        expected_exception: type = Exception,
        name: str = "circuit_breaker",
    ):
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.expected_exception = expected_exception
        self.name = name
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.last_failure_time: Optional[float] = None
        self.success_count = 0
        self._lock = asyncio.Lock()
        self.total_requests = 0
        self.total_failures = 0
        self.total_rejected = 0

    async def call(self, func: Callable, *args, **kwargs) -> Any:
        async with self._lock:
            self.total_requests += 1
            if self.state == CircuitState.OPEN:
                if self._should_attempt_recovery():
                    logger.info("[%s] Attempting recovery -> HALF_OPEN", self.name)
                    self.state = CircuitState.HALF_OPEN
                    self.success_count = 0
                else:
                    self.total_rejected += 1
                    raise CircuitBreakerOpenError(
                        f"Circuit breaker [{self.name}] is OPEN; retry after {self.recovery_timeout:.0f}s"
                    )
        try:
            result = await func(*args, **kwargs)
            async with self._lock:
                if self.state == CircuitState.HALF_OPEN:
                    self.success_count += 1
                    if self.success_count >= 2:
                        logger.info("[%s] Recovery successful -> CLOSED", self.name)
                        self.state = CircuitState.CLOSED
                        self.failure_count = 0
                        self.success_count = 0
                elif self.state == CircuitState.CLOSED:
                    self.failure_count = 0
            return result
        except self.expected_exception as e:
            async with self._lock:
                self.total_failures += 1
                self.last_failure_time = time.time()
                if self.state == CircuitState.HALF_OPEN:
                    self.state = CircuitState.OPEN
                    self.failure_count = self.failure_threshold
                elif self.state == CircuitState.CLOSED:
                    self.failure_count += 1
                    if self.failure_count >= self.failure_threshold:
                        logger.error(
                            "[%s] Threshold %d reached -> OPEN. Error: %s",
                            self.name,
                            self.failure_threshold,
                            e,
                        )
                        self.state = CircuitState.OPEN
            raise

    def _should_attempt_recovery(self) -> bool:
        if self.last_failure_time is None:
            return False
        return (time.time() - self.last_failure_time) >= self.recovery_timeout

    def get_stats(self) -> dict:
        return {
            "name": self.name,
            "state": self.state.value,
            "failure_count": self.failure_count,
            "total_requests": self.total_requests,
            "total_failures": self.total_failures,
            "total_rejected": self.total_rejected,
        }


class CircuitBreakerOpenError(Exception):
    """Raised when circuit is open and request is rejected."""
    pass


# Shared instances for DB and RabbitMQ (plan § 2.9)
db_circuit_breaker = CircuitBreaker(
    failure_threshold=5,
    recovery_timeout=60.0,
    expected_exception=Exception,
    name="db",
)
rabbitmq_circuit_breaker = CircuitBreaker(
    failure_threshold=5,
    recovery_timeout=60.0,
    expected_exception=Exception,
    name="rabbitmq",
)
